bug: Panic with unique composite index on relation with filter on relation

Reported by @david-on-github 

With this schema
```graphql
type Foo {
  id: String
  bars: [Bar] @relation(name: "foobars")
}

type Bar @index(unique: true, includes: [{ name: "foo_id" }, { name: "id" }], name: "test_unqiue") {
  id: String
  foo: Foo @relation(name: "foobars")
}
```

Create some docs for them
Then try to query with a filter on the child relation
```graphql
query {
  Foo {
    _docID
    bars(filter: { foo_id: {_eq: "bae-7eda71ae-2b1c-5ee8-9b21-d8e28b2bf8a3"} }) {
      _docID
      foo_id
    }
  }
}
```
 
Results in this panic.
```
panic: runtime error: invalid memory address or nil pointer dereference
 
 -> github.com/sourcenetwork/defradb/internal/db/fetcher.(*indexPrefixIterator).Close
 ->   github.com/sourcenetwork/defradb/internal/db/fetcher/indexer_iterators.go:144

    github.com/sourcenetwork/defradb/internal/db/fetcher.(*IndexFetcher).Close
      github.com/sourcenetwork/defradb/internal/db/fetcher/indexer.go:215
    github.com/sourcenetwork/defradb/internal/lens.(*lensedFetcher).Close
      github.com/sourcenetwork/defradb/internal/lens/fetcher.go:192
    github.com/sourcenetwork/defradb/internal/planner.(*scanNode).Close
      github.com/sourcenetwork/defradb/internal/planner/scan.go:229
    github.com/sourcenetwork/defradb/internal/planner.(*invertibleTypeJoin).Close
      github.com/sourcenetwork/defradb/internal/planner/type_join.go:491
    github.com/sourcenetwork/defradb/internal/planner.(*typeIndexJoin).Close
      github.com/sourcenetwork/defradb/internal/planner/type_join.go:132
    github.com/sourcenetwork/defradb/internal/planner.(*selectNode).Close
      github.com/sourcenetwork/defradb/internal/planner/select.go:189
    github.com/sourcenetwork/defradb/internal/planner.(*selectTopNode).Close
      github.com/sourcenetwork/defradb/internal/planner/select.go:89
    github.com/sourcenetwork/defradb/internal/planner.(*invertibleTypeJoin).Close
      github.com/sourcenetwork/defradb/internal/planner/type_join.go:495
    github.com/sourcenetwork/defradb/internal/planner.(*typeIndexJoin).Close
      github.com/sourcenetwork/defradb/internal/planner/type_join.go:132
    github.com/sourcenetwork/defradb/internal/planner.(*selectNode).Close
      github.com/sourcenetwork/defradb/internal/planner/select.go:189
    github.com/sourcenetwork/defradb/internal/planner.(*selectTopNode).Close
      github.com/sourcenetwork/defradb/internal/planner/select.go:89
    github.com/sourcenetwork/defradb/internal/planner.(*operationNode).Close
      github.com/sourcenetwork/defradb/internal/planner/operation.go:66
    github.com/sourcenetwork/defradb/internal/planner.(*Planner).RunRequest.func1
      github.com/sourcenetwork/defradb/internal/planner/planner.go:515
    github.com/sourcenetwork/defradb/internal/planner.(*Planner).RunRequest
      github.com/sourcenetwork/defradb/internal/planner/planner.go:545
    github.com/sourcenetwork/defradb/internal/db.(*db).execRequest
      github.com/sourcenetwork/defradb/internal/db/request.go:53
    github.com/sourcenetwork/defradb/internal/db.(*db).ExecRequest
      github.com/sourcenetwork/defradb/internal/db/store.go:38
    github.com/sourcenetwork/defradb/http.(*storeHandler).ExecRequest
      github.com/sourcenetwork/defradb/http/handler_store.go:311
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/chi/v5.(*Mux).routeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:459
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/chi/v5.(*Mux).ServeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:73
    github.com/sourcenetwork/defradb/http.(*Router).ServeHTTP
      github.com/sourcenetwork/defradb/http/router.go:67
    github.com/go-chi/chi/v5.(*Mux).routeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:459
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/sourcenetwork/defradb/http.AuthMiddleware.func1
      github.com/sourcenetwork/defradb/http/auth.go:57
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/sourcenetwork/defradb/http.TransactionMiddleware.func1
      github.com/sourcenetwork/defradb/http/middleware.go:83
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/sourcenetwork/defradb/http.NewHandler.func1.ApiMiddleware.1.1
      github.com/sourcenetwork/defradb/http/middleware.go:71
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/chi/v5.(*Mux).ServeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:73
    github.com/go-chi/chi/v5.(*Mux).Mount.func1
      github.com/go-chi/chi/v5@v5.1.0/mux.go:327
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/chi/v5.(*Mux).routeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:459
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/chi/v5.(*Mux).ServeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:73
    github.com/sourcenetwork/defradb/http.(*Handler).ServeHTTP
      github.com/sourcenetwork/defradb/http/handler.go:107
    github.com/go-chi/chi/v5.(*Mux).routeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:459
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/cors.(*Cors).Handler-fm.(*Cors).Handler.func1
      github.com/go-chi/cors@v1.2.1/cors.go:228
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/chi/v5/middleware.Recoverer.func1
      github.com/go-chi/chi/v5@v5.1.0/middleware/recoverer.go:45
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/sourcenetwork/defradb/http.NewServer.RequestLogger.func2.1
      github.com/go-chi/chi/v5@v5.1.0/middleware/logger.go:55
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171
    github.com/go-chi/chi/v5.(*Mux).ServeHTTP
      github.com/go-chi/chi/v5@v5.1.0/mux.go:90
    net/http.serverHandler.ServeHTTP
      net/http/server.go:3142
    net/http.(*conn).serve
      net/http/server.go:2044
    created by net/http.(*Server).Serve in goroutine 75
      net/http/server.go:3290
```

--- a/internal/db/fetcher/indexer_iterators.go
+++ b/internal/db/fetcher/indexer_iterators.go
@@ -141,6 +141,9 @@ func (iter *indexPrefixIterator) Next() (indexIterResult, error) {
 }
 
 func (iter *indexPrefixIterator) Close() error {
+	if iter.resultIter == nil {
+		return nil
+	}
 	return iter.resultIter.Close()
 }
 

--- a/tests/integration/index/query_with_relation_filter_test.go
+++ b/tests/integration/index/query_with_relation_filter_test.go
@@ -1018,36 +1018,36 @@ func TestQueryWithIndexOnManyToOne_MultipleViaOneToMany(t *testing.T) {
 	testUtils.ExecuteTestCase(t, test)
 }
 
-func TestQueryWithIndex_UniqueIndexOnChildWithEmptyParentCollection(t *testing.T) {
+func TestQueryWithUniqueIndex_WithFilterOnChildIndexedField_ShouldFetch(t *testing.T) {
 	test := testUtils.TestCase{
 		Actions: []any{
 			testUtils.SchemaUpdate{
 				Schema: `
-					type Action {
-						key: String @index(unique: true)
-						playerActions: [PlayerAction]
+					type User {
+						name: String @index(unique: true)
+						devices: [Device]
 					}
 
-					type PlayerAction {
-						deleted: Boolean
-						action: Action
+					type Device {
+						trusted: Boolean
+						owner: User
 					}
 				`,
 			},
 			testUtils.CreateDoc{
 				CollectionID: 0,
 				DocMap: map[string]any{
-					"key": "ACTION_KEY",
+					"name": "John",
 				},
 			},
 			testUtils.Request{
 				Request: `query {
-					PlayerAction(filter: {action: {key: {_eq: "ACTION_KEY"}}}) {
-						deleted
+					Device(filter: {owner: {name: {_eq: "John"}}}) {
+						trusted
 					}
 				}`,
 				Results: map[string]any{
-					"PlayerAction": []map[string]any{},
+					"Device": []map[string]any{},
 				},
 			},
 		},

--- a/tests/integration/index/query_with_unique_index_on_relation_filter_test.go
+++ b/tests/integration/index/query_with_unique_index_on_relation_filter_test.go
@@ -1,0 +1,72 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package index
+
+import (
+	"testing"
+
+	testUtils "github.com/sourcenetwork/defradb/tests/integration"
+)
+
+func TestQueryWithUniqueCompositeIndex_WithFilterOnIndexedRelation_ShouldFilter(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: `
+					type User {
+						name: String 
+						devices: [Device]
+					}
+
+					type Device  {
+						manufacturer: String 
+						owner: User @index(unique: true, includes: [{name: "manufacturer"}])
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				CollectionID: 0,
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.CreateDoc{
+				CollectionID: 1,
+				DocMap: map[string]any{
+					"manufacturer": "Apple",
+					"owner_id":     testUtils.NewDocIndex(0, 0),
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						name
+						devices(filter: {owner_id: {_eq: "bae-1ef746f8-821e-586f-99b2-4cb1fb9b782f"}}) {
+							manufacturer
+						}
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"name": "John",
+							"devices": []map[string]any{
+								{"manufacturer": "Apple"},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}

--- a/tests/integration/index/query_with_unique_index_on_relation_filter_test.go
+++ b/tests/integration/index/query_with_unique_index_on_relation_filter_test.go
@@ -1,4 +1,4 @@
-// Copyright 2023 Democratized Data Foundation
+// Copyright 2024 Democratized Data Foundation
 //
 // Use of this software is governed by the Business Source License
 // included in the file licenses/BSL.txt.

